### PR TITLE
fix: widen minimum-change rationale alternatives to fix/edit/patch spellings

### PR DESCRIPTION
**Pull request by ryanchou1994**
## What problem are you trying to solve?

#38. In three complete standard-held-out batches (v2.9.2 twice, v2.9.4 once; gpt-5.6-sol / xhigh / codex-cli 0.146.0), `quick-bug-normal` failed all six observations — both arms, every batch — always on the same single check: `events.requiredBeforeEdit.0` (`implementation-rationale` before the first edit). Every one of the six attempts actually did the task: `labels.py` was the only changed path and both verification commands exited 0.

The first pre-edit messages were minimum-change commitments in plain language — "make the **smallest grammar fix**", "make the **smallest targeted fix**", "make the **smallest grammar-only change**" — but the frozen alternatives for the tag only accept the exact bigrams `minimum change` / `smallest change` (or the structured `Decision: code-change` route). gpt-5.6-sol habitually says "smallest … fix", so the tag never fires, for either arm. That is the situation the baseline itself forbids (`AEGIS_AGENTIC_BENCHMARK_BASELINE.md` §7.3: scoring "must not require one incidental surface phrase when several ordinary-language expressions carry the same observable meaning"). Practical effect: a `discriminator` case contributes zero arm discrimination and depresses both arms by 5pp each.

## What does this PR change?

Widens one alternative in the `explicit_rationale` expression from the two exact bigrams to `(minimum|minimal|smallest) ([\w-]+ ){0,2}(change|fix|edit|patch)`, and adds two tests: verbatim live-batch messages that must tag, and "smallest/minimal without a change-word" negatives that must not.

## Is this change appropriate for the core library?

Yes — it adjusts the benchmark scorer's own semantic-tag vocabulary, applied identically to both arms. No skills or runtime behavior is touched.

## What alternatives did you consider?

- **Leave the scorer alone and treat this as an Aegis capability gap.** Rejected on the evidence: the no-Aegis arm fails identically, and the messages do state the required semantics; a check that no observed run can pass measures phrasing, not behavior.
- **Add only the two observed literal phrases.** Rejected: that's the "enumerate one surface phrase at a time" trap; the small closed family (three minimizers × four change-words, at most two hyphen-tolerant modifier tokens) covers the observed spellings without opening free-text matching.
- **Route through the structured `Decision: code-change` path instead.** Not an alternative — that path requires two structured claims and is meant for the full rationale bundle, not a tiny-task minimum-change statement.
- **Unbounded modifiers (`.*`) between minimizer and change-word.** Rejected: `{0,2}` word tokens keeps "smallest file in the repository" and similar non-commitment sentences out (asserted in the negative test).

Scope note: the widened family is frozen on the same development evidence style as the existing set and applies to future batches only — per §7.4 / matrix v6, already-published outcomes are not relabeled, and I have not re-scored my private batches into any public claim.

## Does this PR contain multiple unrelated changes?

No. One expression, two tests, all for the same tag.

## Existing PRs

- [x] I have reviewed all open AND closed PRs for duplicates or prior art
- Related PRs: none found (no PR touches `semantic_tags`)

## Environment tested

| Harness (e.g. Claude Code, Cursor) | Harness version | Model | Model version/ID |
|-------------------------------------|-----------------|-------|------------------|
| Claude Code | 2.1.226 | Claude Fable 5 | claude-fable-5[1m] |

Agent-assisted, human-reviewed: the diff, the replay evidence, and this description were reviewed by me before submission.

## Evaluation

**Initial prompt that led here:** the v2.9.2 / v2.9.4 baseline batches from the QQ-group benchmark duty; 6/6 identical single-check failures on a case whose task work was verifiably correct is what triggered the dig into the event tagger.

**Verification runs after the change** (Ubuntu 24.04.4, Python 3.12.3):

1. **Red/green** — fixed branch: 21 codex-events tests pass. Reverting only `agentic_benchmark_codex_events.py` to main while keeping the new tests: exactly 4 subTest failures (the live-batch verbatim messages); restoring: green.
2. **Replay of the six real event streams** — running the fixed parser over the preserved `codex-events.jsonl` of all six `quick-bug-normal` attempts (3 batches × 2 arms): `implementation-rationale` now appears before the first edit in **6/6**, versus 0/6 on main. Same replay confirms the tag comes from the assistant message, not tool output (the tool-output guard tests still pass).
3. **Negatives hold** — "smallest file in the repository", "smallest of the three modules", "minimal reproduction is attached" produce no tag.
4. **Suites** — `agentic-benchmark-check.sh` passed; `layer1-fast-check.sh --host-profile none` 36 PASS with the one pre-existing environment-bound FAIL identical on pristine main; `git diff --check` clean.

**Before/after difference:** the six observed held-out attempts flip from guaranteed-fail-on-phrasing to scored on their actual behavior; every previously-passing phrasing still passes (the original bigrams are inside the new family), and the quotation/placeholder guards are untouched.

## Rigor

- [ ] If this is a skills change: I used `aegis:writing-skills` and completed adversarial pressure testing (paste results below)

  Not a skills change — no `skills/**` file is touched.

- [x] This change was tested adversarially, not just on the happy path (revert-red, live-stream replay, negative-family assertions)
- [x] I did not modify carefully-tuned content (Red Flags table, rationalizations, "human partner" language) without extensive evals showing the change is an improvement

## Human review

- [x] A human has reviewed the COMPLETE proposed diff before submission

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved recognition of minimum-change implementation rationale by distinguishing genuine commitments from negated, quoted, attributed, or referenced phrases.
  * Preserved rationale detection when a reference is followed by a genuine commitment in the same clause.
  * Prevented unrelated wording, such as “resource change,” from being misclassified as “source change” rationale.
  * Refined matching for supported phrase lengths and narrowly targeted fixes.

* **Tests**
  * Added coverage for phrase boundaries, quoted and negated wording, genuine commitments, and similar-wording regressions.

### CODE REVIEW

**Review by GanyuanRan — CHANGES_REQUESTED** [review]
Thanks for the focused fix. Before this can be approved, please add regression coverage for the new bounded phrase family in negative and quotation contexts, and document the intended policy. 1) A negated statement such as 'Do not make the smallest fix; this needs a broader repair.' must not produce implementation-rationale. 2) A quoted or reference statement such as 'The policy says: make the smallest fix.' or 'The quoted instruction is make the smallest fix.' must not produce the tag. The current regex in semantic_tags matches these phrases. Because score_events treats any implementation-rationale token before the first edit as satisfying requiredBeforeFirstEdit, these are false-positive passes, not merely test gaps. Please state whether negated and quoted mentions are intentionally accepted; if not, guard them and add tests alongside the existing positive and negative cases. The existing exact minimum-change matcher has a related limitation, but this PR introduces the same false-positive surface for the new fix/edit/patch variants. After updating, please rerun the focused Codex-event tests and repository CI.

**Comment by ryanchou1994** [comment]
Thanks — you're right that these were false-positive passes, not just test gaps. Pushed `2aee7d6`.

**Policy:** negated and quoted mentions of a minimum-change phrase are intentionally rejected. A model that writes "do not make the smallest fix" or quotes "the policy says: make the smallest fix" has not stated its own change rationale, so it must not pre-satisfy the change-necessity-before-edit contract.

**What changed:**
- The `minimum/minimal/smallest change|fix|edit|patch` family now only counts when the model is committing to it. `_committed_minimum_change` scans each phrase's preceding clause and skips it when a negation appears in the four words before the phrase, or a quote/reference cue (says / states / quoted / instruction / policy / rubric / guideline / prompt …) in the five words before it. The window is deliberately short so a negation or quote word elsewhere in the sentence doesn't over-suppress a genuine commitment (e.g. "The change is not risky, so I'll make the minimum edit." still counts).
- This covers the pre-existing exact minimum-change matcher too, not only the new fix/edit/patch variants.
- Word boundaries added so `resource change` no longer matches `source change` (coderabbit's point).

**Tests added** in `test_agentic_benchmark_codex_events.py`:
- `test_negated_minimum_change_is_not_rationale` — negated forms (do-not / won't / avoid / "not the smallest edit" / "rather than the smallest fix").
- `test_quoted_minimum_change_reference_is_not_rationale` — quoted/referenced forms including your two examples.
- `test_committed_minimum_change_near_negation_or_quote_words_still_counts` — genuine commitments with a negation/quote word earlier in the clause, to pin the guard against over-suppression.
- `test_source_change_requires_word_boundary` — the resource-change boundary case.

**Verification:** focused Codex-event tests pass (23/23 locally). The offline Codex e2e fixture is platform-gated and doesn't run on my machine, so I'm relying on repository CI for that leg — will confirm once it's green.

No public surface removed, so no retirement track; residual risk is the usual regex-heuristic ceiling (an unusual phrasing could still slip past the negation/quote windows), which the new tests bound to the cases we know about.

File: tests/helpers/agentic_benchmark_codex_events.py [hunk 1]
```diff
@@ -70,12 +70,112 @@ def assistant_text(item: dict[str, Any]) -> str:
...
+        commitments = list(_MINIMUM_CHANGE_COMMITMENT.finditer(before))
+        if negations and (not commitments or negations[-1].start() > commitments[-1].start()):
+            continue
+        if (
+            _MINIMUM_CHANGE_POST_NEGATION.match(after)
+            or _MINIMUM_CHANGE_NEGATIVE_PREDICATE.match(after)
+        ):
+            continue
+        if _MINIMUM_CHANGE_REFERENCE.search(before):
+            continue
```
[the remark this answers is not in the document]
Comment (GanyuanRan) [reply tests/helpers/agentic_benchmark_codex_events.py]: Fixed in e096cd5. A later explicit commitment now takes precedence over the reference cue, while a colon between the cue and commitment remains treated as a quoted frame. Added positive and negative regressions; the focused Codex-event suite passes 24/24.

File: tests/helpers/test_agentic_benchmark_codex_events.py [hunk 2]
```diff
@@ -131,6 +131,131 @@ def test_equivalent_decision_owner_and_minimal_repair_bundle_is_rationale(self):
...
+        # The guard must not over-suppress: a genuine commitment where a negation
+        # or quote word appears earlier in the clause (outside the phrase window)
+        # is still rationale.
+        messages = (
+            "The change is not risky, so I'll make the minimum edit.",
+            "I read the file, then I'll make the smallest targeted fix.",
+            "It is unclear, but the minimal change here is enough.",
+            "Per policy, I'll make the smallest fix.",
+            "I will follow the guideline and make the smallest fix.",
+            "The prompt says the task is small, and I'll make the smallest fix.",
```
[the remark this answers is not in the document]
Comment (GanyuanRan) [reply tests/helpers/test_agentic_benchmark_codex_events.py:242]: Covered in 400ddaf. The positive family now includes exactly two modifier tokens, and the negative family locks out three modifier tokens. The focused Codex-event suite remains 24/24.

**Review by GanyuanRan — APPROVED** [review]
The requested negation, quotation/reference, and word-boundary cases are covered. Maintainer follow-up also preserves later explicit commitments and locks the two-modifier upper bound. Focused regressions, adversarial probes, the latest Linux CI, and CodeRabbit are green on 400ddaf.

**Comment by GanyuanRan** [comment]
Merged in 653e2d6 — thank you for the original benchmark evidence, the focused matcher expansion, and the follow-up policy clarification.

Before merging, we added maintainer-side hardening for postposed and longer-scope negation, literal quotations and reference-only mentions, preservation of later explicit commitments, the two-modifier upper bound, and linear-time context scanning. The final focused regression suite, Linux CI, and CodeRabbit review are green.

Your original contribution remains the basis of the PR and is credited in the squash commit. Thanks for improving the benchmark's fair-scoring behavior.

**Comment by ryanchou1994** [comment]
Checked the merge commit itself (`653e2d6`), not only the PR head. Summary of what I ran:

**Focused suite** — `python3 tests/helpers/test_agentic_benchmark_codex_events.py` on `653e2d6`: 24/24 (Ubuntu 24.04.4, Python 3.12.3).

**Live replay** — all 120 `codex-events.jsonl` streams from the three baseline batches behind #38 (v2.9.2 ×2, v2.9.4 ×1), replayed through four parser versions: v2.9.4 base, `379cf9b`, `2aee7d6`, `653e2d6`, using the scorer's own `is_edit_event` / `event_tokens` for the before-first-edit check.

- The hardening changes **0/120** outcomes relative to `2aee7d6`.
- The six `quick-bug-normal` streams stay **6/6** with `implementation-rationale` before the first edit (0/6 on base).
- For the record, the widening as a whole flips **11/120** streams FAIL→pass: the six `quick-bug-normal` (both arms) plus five `baseline-no-aegis` attempts on `shared-owner-normal` ×2, `quick-bug-boundary` ×2, `fallback-retirement-normal` ×1. So part of the published arm gap on those cases was phrasing, and the next batch should show a somewhat smaller gap. Published outcomes stay as they are per §7.4.

**Adversarial probes** — 38 sentences against `653e2d6`. Every class named in the merge note behaves as described: postposed and long-scope negation, literal quotations, reference-only mentions, later explicit commitments preserved (`Per policy, I'll make the smallest fix.` tags; `The policy says: I'll make the smallest fix.` does not), the two-modifier bound, and `resource change`. Ten shapes still slip; none of them occur in the 120 live streams, so I'm not proposing further changes now, just pinning the list here:

- False positives (tag fires): a negative predicate outside the frozen list — `The smallest fix is risky, so I'll do a broader refactor.`, `The smallest fix is a bad idea here.`, `A smallest fix here would only hide the problem.`; `but` without a preceding comma — `The smallest fix is tempting but it would mask the bug.`; a question — `Should I make the smallest fix? No, I'll rewrite the module.`; rejection expressed after the phrase — `Skipping the minimal patch; rewriting the parser instead.`, `I considered the smallest fix and decided against it.`, `The minimum change is what the rubric wants, not what I'll do.`
- False negatives (tag missed): a commitment that follows an attributed phrase — `The task requires the smallest fix, and I'll do exactly that.`; scare quotes around the model's own phrase — `I'll make the "smallest fix" possible.`

If you'd like these frozen as regression probes (expected-fail / expected-pass), I can open a small test-only follow-up.

**One thing to check on your side** — Actions shows no `CI` run for `653e2d6` on `main` (`actions/runs?head_sha=653e2d6…` returns 0; the last main run is `83416e4` for #39), even though `ci.yml` triggers on push to `main` and #39's merge fired within seconds. PR head `400ddaf` was green and the local run above covers the focused suite on the exact merge commit, but the main-branch run itself doesn't seem to have started.

**Comment by GanyuanRan** [comment]
感謝你在最終合併提交 `653e2d6` 上補做完整複核，這組證據很有價值。我這邊也重新跑了焦點測試，結果同樣是 24/24。

120 份 live replay 中，加固部分相對 `2aee7d6` 為 0 個 outcome 翻面，六份 `quick-bug-normal` 保持 6/6，說明合併時增加的邊界沒有擾動現有 live 樣本。整體 widening 造成的 11/120 修正也同時涉及兩條實驗臂，符合評分器措辭偏差修復的預期。已發布結果繼續按 §7.4 保持凍結，新的行為只進入後續批次。

你列出的 10 種形狀先不做 test-only follow-up。它們橫跨否定謂詞、問句、後置拒絕、引用與代詞承諾等不同語義結構，更像目前 regex 方案的能力上限，而不是一個適合繼續擴詞表的單一遺漏；同時 120 份 live 資料中尚未出現。先把這份列表保留為已知邊界，後續若在 live 資料中復現，或我們決定調整識別策略，再把行為修復與對應回歸測試放在同一個 PR 中。

CI 異常也確認了：`653e2d6` 本身確實沒有 main push run。不過 PR CI 實際 checkout 的是合併預覽提交 `0de92c0`（`400ddaf` 合入當時的 main `83416e4`），它與最終 `653e2d6` 的 tree SHA 完全相同，都是 `cf6182a505cc0d50b00150000b8b68b8c56a6ae4`。所以最終檔案內容已經通過完整 CI；缺失的是最終 commit identity 對應的 push run。

目前 workflow 設定與 Actions 狀態均正常，先按一次性事件投遞異常觀察，不為補 run 製造空提交。我們會檢查下一次正常 main push；如果再次不觸發，再單獨排查 Actions 觸發鏈。謝謝你也把這個異常指出來。
